New readers from MM satellite group

rrbuchholz opened the pull request:
Adding in:
Tropomi L2 reader
OMPS L3 reader
MOPITT L3 reader
and adding to `__init__.py` file

rrbuchholz committed:
Develop satellite (#29)

* Update raqms.py

tracking problems in colab

* Update __init__.py

* Update __init__.py

* Update __init__.py

* trying to make Kaggle import work

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update setup.cfg

* Update setup.cfg

* Update __init__.py

* Update raqms.py

adjusting timestamp. Needed for pairing. Unsure if this was a temporary fix sort of thing

* Update raqms.py

* Update raqms.py

* Update raqms.py

* Update omps_nadir.py

* Update omps_nadir.py

* data handling issue fix

* Update omps_nadir.py

* Update omps_nadir.py

* add trimmed file reader for earthcube notebook

* Update raqms.py

* Update raqms.py

* Update raqms.py

* merge fixes

* trying to fix for merge

* Adding MOPITT read

* tropomi no2 reader

* rename

* adding hdf close file

* omps level 3 reader

* Update raqms.py

remove double of _fix_pres

---------

Co-authored-by: mebruckner <48494069+mebruckner@users.noreply.github.com>
Co-authored-by: Maggie Bruckner <mbruckner@raqms-ops.ssec.wisc.edu>
Co-authored-by: Meng Li <mengli@MengM1.local>

zmoon commented:
@dwfncar @rrbuchholz do you have/want any tests for any of these?

zmoon commented:
3.6 failing due to #114 , but there are also some comments from flake8 to address, and some formatting stuff.
Also looks like we need to add h5py to the docs env.

rrbuchholz commented:
I have an example Jupyter notebook for MOPITT in the MELODIES MONET develop_satellite branch that needs to be pulled.
I also think Meng has added a Tropomi example and Maggie has added the OMPS L3 example in MELODIES MONET.

zmoon commented:
@rrbuchholz that'll be great to have those MM examples, but I was referring here to smaller test cases for the MONETIO readers to add to the MONETIO test suite. It's ok if we don't have, just thought I would ask.

zmoon committed:
format

zmoon committed:
please flake8

zmoon committed:
No top-level h5py import; some formatting

zmoon commented:
Data file @rrbuchholz shared now available at https://csl.noaa.gov/groups/csl4/modeldata/melodies-monet/data/example_observation_data/satellite/MOP03JM-201701-L3V95.9.3.he5

zmoon committed:
Initial mop test

zmoon committed:
Test da extractor

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+
+    try:
+        inFile = h5py.File(filename, "r")
+    except Exception:
+        print("ERROR: CANNOT OPEN " + filename)
+        return 0
```
Fine to just raise exception here.

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+    import h5py
+
+    structure = "/HDFEOS/ADDITIONAL/FILE_ATTRIBUTES"
+    # print("READING FILE " + inFileName)
+    # fName = os.path.basename(filename)
```
```suggestion
```

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+    k = grp.attrs
+    startTimeBytes = k.get("StartTime", default=None)
+    startTime = pd.to_datetime(startTimeBytes[0], unit="s", origin="1993-01-01 00:00:00")
+    # print("******************", startTime)
```
```suggestion
```

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+
+    try:
+        inFile.close()
+    except Exception:
+        print("ERROR CANNOT CLOSE " + filename)
+        return 0
```
Fine to just let possible exception be raised here.

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+    }
+    try:
+        data_loaded = he5_load[variable_dict[varname]][:]
+    except Exception:
+        print("ERROR: Cannot load " + varname + " from " + filename)
+        return 0
```
Fine to just let possible exception be raised here.

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+            data_loaded, dims=["lon", "lat", "alt"], coords=[lon, lat, alt_short]
+        )
+
+    return ds
```
How about at least adding `units` and `long_name` to the variables?

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+    Parameters
+    ----------
+    files : str or list of str
+        The full path to the file or files. Can take a file template with wildcard (*) symbol.
```
Currently list of strings will fail.

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+        data = loadAndExtractGriddedHDF(filename, varname)
+        time = getStartTime(filename)
+        data = data.expand_dims(axis=0, time=[time])
+        if count == 0:
```
Currently count is always 0 when we get here the first time.

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+    return ds
+
+
+def read_mopittdataset(files, varname):
```
```suggestion
def open_dataset(files, varname):
```
More general name. The module it is in still identifies it to be for MOPITT.

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+            full_dataset = data
+            count += 1
+        else:
+            full_dataset = xr.concat([full_dataset, data], "time")
```
Could do the concat after the load loop.

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+import xarray as xr
+
+
+def getStartTime(filename):
```
```suggestion
def get_start_time(filename):
```
Python naming style

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,147 @@
+    return startTime
+
+
+def loadAndExtractGriddedHDF(filename, varname):
```
```suggestion
def _open_single_dataset(filename, varname):
```
Seems like the intention is not to use this directly.

zmoon commented on monetio/sat/_omps_l3_mm.py:
```diff
@@ -0,0 +1,72 @@
+def read_OMPS_l3(files):
+    """Loop to open OMPS nadir mapper Total Column Ozone L3 files.
+    Parameters:
+    -----------
+    files: string or list of strings
+
+    returns xarray Dataset.
```
```suggestion

    Parameters
    ----------
    files: str or list of str

    Returns
    -------
    xarray.Dataset
```

zmoon commented on monetio/sat/_omps_l3_mm.py:
```diff
@@ -0,0 +1,72 @@
+def read_OMPS_l3(files):
```
Could use a general name like `open_dataset` here (if applicable).

rrbuchholz replied:
Each satellite will likely require different opening methodology so I think we need to specify in the naming what type of file reading is occurring.

zmoon replied:
Module scoping does allow the names to be the same, e.g. the MM model reader functions are called simply `open_mfdataset`.

zmoon commented on monetio/sat/_omps_l3_mm.py:
```diff
@@ -0,0 +1,72 @@
+
+def extract_OMPS_l3(fname):
+    """Read locally stored NASA Suomi NPP OMPS Level 3 Nadir Mapper TO3 files
+    Parameters
+    ----------
+    fname: string
+        fname is local path to h5 file
+
+    Returns
+    -------
+    ds: xarray dataset
+
```
```suggestion

    Parameters
    ----------
    fname: str
        Local path to h5 file.

    Returns
    -------
    xarray.Dataset
```

zmoon commented on monetio/sat/_tropomi_l2_no2_mm.py:
```diff
@@ -0,0 +1,138 @@
+def read_dataset(fname, variable_dict):
+    """
+    Parameters
+    __________
```
```suggestion
    ----------
```
And the other ones in this module too.

zmoon reviewed (commented):
Some initial comments.

zmoon committed:
better msg

zmoon committed:
Add h5py to dev env

rrbuchholz committed:
Merge branch 'noaa-oar-arl:develop' into develop

rrbuchholz committed:
Update monetio/sat/_tropomi_l2_no2_mm.py

text formatting differences

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

rrbuchholz committed:
Update monetio/sat/_omps_l3_mm.py

text formatting using consistent language

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

rrbuchholz committed:
Update monetio/sat/_omps_l3_mm.py

using consistent language in text descriptions

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

rrbuchholz committed:
Update monetio/sat/_mopitt_l3_mm.py

Code semantics

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

rrbuchholz committed:
Update monetio/sat/_mopitt_l3_mm.py

Remove development debugging

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

rrbuchholz committed:
Update monetio/sat/_mopitt_l3_mm.py

Remove development debugging

Co-authored-by: Zachary Moon <zmoon92@gmail.com>

rrbuchholz committed:
Adding units and long_name

zmoon commented on monetio/sat/_mopitt_l3_mm.py:
```diff
@@ -0,0 +1,142 @@
+    # print(data_loaded.shape)
+
+    # create xarray DataArray
+    if varname == "column":
```
I see we are not covering all cases that are in `variable_dict`

zmoon replied:
I think f097e22 took care of this

zmoon committed:
Style; fix file list/str

zmoon committed:
Test MOPITT `open_dataset` too

zmoon committed:
Let exceptions be raised; check `varname`

with (somewhat) informative message containing the options

zmoon commented on monetio/sat/_tropomi_l2_no2_mm.py:
```diff
@@ -0,0 +1,138 @@
+        ds[varname] = xr.DataArray(values)
+
+        if "quality_flag_min" in variable_dict[varname]:
+            ds.attrs["quality_flag"] = varname
+            ds.attrs["quality_thresh_min"] = variable_dict[varname]["quality_flag_min"]
```
These are dataset attrs so they will be overwritten if there is more than one variable in `variable_dict`.

mlirenzhenmayi replied:
True. This is an old version of the TROPOMI reader. I add a new dataset attribute 'var_applied' to define the variables list for each quality flag in the new reader

zmoon replied:
@mlirenzhenmayi I don't see this in the current version of the code.

zmoon commented on monetio/sat/_tropomi_l2_no2_mm.py:
```diff
@@ -0,0 +1,138 @@
+                values[quality_flag <= quality_thresh_min] = np.nan
+
+
+def open_dataset(fnames, variable_dict, debug=False):
```
The `variable_dict` is necessary for the fn to be useful. Is this something that will always be the same? If so, could include in the code sort of like in the MOPITT one.

and/or document in the docstring the format

mlirenzhenmayi replied:
fn always change with the date and vary one by one. It seems that there's no string always the same.

zmoon replied:
By the way, by "fn" I meant function, not filename. I was wondering if we could set it up so you could pass `varnames` like the MOPITT one, instead of the variable dict. The variable dict info could be embedded in the module, if it is fixed for a certain variable.

zmoon committed:
File lock for MOPITT test file

zmoon committed:
Add filelock to dev deps

zmoon committed:
Alternate lock method

avoiding the copying

difference is that in the xdist case the file is already downloaded
before the lock context manager entered

zmoon committed:
Revert somewhat

zmoon committed:
Merge remote-tracking branch 'noaa/develop' into ncar-develop

zmoon committed:
cleanup

rrbuchholz committed:
Merge branch 'noaa-oar-arl:develop' into develop

rrbuchholz committed:
Merge branch 'noaa-oar-arl:develop' into develop

mebruckner committed:
Modifications to Mopitt L3 reader  (#31)

* Update raqms.py

tracking problems in colab

* Update __init__.py

* Update __init__.py

* Update __init__.py

* trying to make Kaggle import work

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update __init__.py

* Update setup.cfg

* Update setup.cfg

* Update __init__.py

* Update raqms.py

adjusting timestamp. Needed for pairing. Unsure if this was a temporary fix sort of thing

* Update raqms.py

* Update raqms.py

* Update raqms.py

* Update omps_nadir.py

* Update omps_nadir.py

* data handling issue fix

* Update omps_nadir.py

* Update omps_nadir.py

* add trimmed file reader for earthcube notebook

* Update raqms.py

* Update raqms.py

* Update raqms.py

* merge fixes

* trying to fix for merge

* Adding MOPITT read

* tropomi no2 reader

* rename

* adding hdf close file

* omps level 3 reader

* Update raqms.py

remove double of _fix_pres

* personal changes for using wisc wrf-chem output

* Update _mopitt_l3_mm.py

updating to read multiple variables in rather than just 1

* Update _mopitt_l3_mm.py

fixing small error

* Update _mopitt_l3_mm.py

Removing loading of multiple variables in read_mopittdataset to figure out why apriori_col isn't loading right

* Update _mopitt_l3_mm.py

Fix consistency error in naming for apriori_prof, adding  back in initial update for reading in many variables from file in one go

* correct merge conflicts

* add updated version to correct merge conflicts

* fix merge conflicts

* finish merge conflict fixes

* typo

* forgot a comma

* fix merge conflicts by copying newer version of reader in

* updated mopitt reader to add all variables needed for pairing and column calculation

* remove reference to a reader not included

Removed call to a reader that is only in personal branch

---------

Co-authored-by: R Buchholz <rrbuchholz@users.noreply.github.com>
Co-authored-by: Maggie Bruckner <mbruckner@raqms-ops.ssec.wisc.edu>
Co-authored-by: rrbuchholz <buchholz@ucar.edu>
Co-authored-by: Meng Li <mengli@MengM1.local>

rrbuchholz committed:
Merge branch 'noaa-oar-arl:develop' into develop

rrbuchholz commented:
Hi @zmoon! Just wondering what you need from us in the satellite team to get this pull request merged? Would be great to try and have it done before the summer release of MELODIES MONET. (It's been a year since we opened this!)

zmoon commented:
Hi @rrbuchholz , very sorry, it's been on my list to get back to @mlirenzhenmayi

zmoon committed:
Merge branch 'develop' into develop

zmoon committed:
Auto format

zmoon committed:
No ##

flake8 doesn't like them

zmoon committed:
Allow single string varname

this way the current test still works

zmoon committed:
Variable attrs instead

seems to be more likely what we want

though I'm not sure why setting 'quality_flag' to the var name
is helpful

zmoon committed:
Sanity check

zmoon committed:
sp

zmoon committed:
Allow passing Path

zmoon committed:
dims and such; get time from data

zmoon committed:
Initial tropomi test

zmoon committed:
Apply dataset fill value

zmoon committed:
Check min thing works

zmoon committed:
Only modify logging config in debug case

zmoon committed:
Update docstring

zmoon committed:
Switch y and x

zmoon committed:
Add attrs; time as coord

zmoon committed:
Initial OMPS test using local files

zmoon committed:
Use uploaded data

zmoon commented:
@rrbuchholz and @mlirenzhenmayi I added some basic tests for the other two. I also made a few changes here and there (e.g. adding attributes to variables), maybe you can check that your workflows still work and/or review.

zmoon committed:
Back to single quality flag var

applied to all data vars

zmoon committed:
Allow passing vn(s); doc variable_dict a bit

rrbuchholz committed:
Merge branch 'noaa-oar-arl:develop' into develop

rrbuchholz commented:
It all looks good to merge.

zmoon commented:
@rrbuchholz thanks for checking it out. I haven't added @mlirenzhenmayi 's updates to tropomi but maybe it's simpler to merge this now and add that after.

rrbuchholz commented:
@zmoon Yeah, I think that sounds like a good idea.
